### PR TITLE
[SPARK-14004][SQL] NamedExpressions should have at most one qualifier

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

This is a more aggressive version of PR #11820, which not only fixes the original problem, but also does the following updates to enforce the at-most-one-qualifier constraint:
- Renames `NamedExpression.qualifiers` to `NamedExpression.qualifier`
- Uses `Option[String]` rather than `Seq[String]` for `NamedExpression.qualifier`

Quoted PR description of #11820 here:

> Current implementations of `AttributeReference.sql` and `Alias.sql` joins all available qualifiers, which is logically wrong. But this implementation mistake doesn't cause any real SQL generation bugs though, since there is always at most one qualifier for any given `AttributeReference` or `Alias`.
## How was this patch tested?

Existing tests should be enough.

### CODE REVIEW

**Comment by liancheng** [comment]
Personally, I had once been quite confused by the fact that `NamedExpression.qualifiers` is a `Seq[String]` and thought that attributes can be qualified with multiple qualifiers like `db.table.column`. That's why current version of `AttributeReference.sql` joins all qualifiers using `.` rather than picking the first one.

I believe it's safe and good to enforce the at-most-one-qualifier constraint at type level unless there do exist valid cases where using multiple qualifiers makes sense but haven't been implemented in Spark SQL yet.

cc @marmbrus @rxin @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1438,7 +1438,7 @@ object CleanupAliases extends Rule[LogicalPlan] {
   def trimNonTopLevelAliases(e: Expression): Expression = e match {
     case a: Alias =>
       Alias(trimAliases(a.child), a.name)(
-        a.exprId, a.qualifiers, a.explicitMetadata, a.isGenerated)
+        a.exprId, a.qualifier, a.explicitMetadata, a.isGenerated)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: not related to this PR, but we can simplify this to: `a.withNewChildren(trimAliases(a.child) :: Nil)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 2]
```diff
@@ -122,17 +122,17 @@ abstract class Attribute extends LeafExpression with NamedExpression {
  * @param name The name to be associated with the result of computing [[child]].
  * @param exprId A globally unique id used to check if an [[AttributeReference]] refers to this
  *               alias. Auto-assigned if left blank.
- * @param qualifiers A list of strings that can be used to referred to this attribute in a fully
+ * @param qualifier A list of strings that can be used to referred to this attribute in a fully
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: we need to update the comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 3]
```diff
@@ -122,17 +122,17 @@ abstract class Attribute extends LeafExpression with NamedExpression {
...
  *                   tableName and subQueryAlias are possible qualifiers.
  * @param explicitMetadata Explicit metadata associated with this alias that overwrites child's.
  * @param isGenerated A flag to indicate if this alias is generated by Catalyst
  */
 case class Alias(child: Expression, name: String)(
-    val exprId: ExprId = NamedExpression.newExprId,
-    val qualifiers: Seq[String] = Nil,
-    val explicitMetadata: Option[Metadata] = None,
-    override val isGenerated: java.lang.Boolean = false)
+  val exprId: ExprId = NamedExpression.newExprId,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: wrong ident?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 4]
```diff
@@ -197,9 +197,9 @@ case class Alias(child: Expression, name: String)(
  * @param metadata The metadata of this attribute.
  * @param exprId A globally unique id used to check if different AttributeReferences refer to the
  *               same attribute.
- * @param qualifiers A list of strings that can be used to referred to this attribute in a fully
- *                   qualified way. Consider the examples tableName.name, subQueryAlias.name.
- *                   tableName and subQueryAlias are possible qualifiers.
+ * @param qualifier A list of strings that can be used to referred to this attribute in a fully
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: update the comment here too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 5]
```diff
@@ -264,18 +264,18 @@ case class AttributeReference(
...
   }
 
   /**
-   * Returns a copy of this [[AttributeReference]] with new qualifiers.
+   * Returns a copy of this [[AttributeReference]] with new qualifier.
    */
-  override def withQualifiers(newQualifiers: Seq[String]): AttributeReference = {
-    if (newQualifiers.toSet == qualifiers.toSet) {
+  override def withQualifier(newQualifier: Option[String]): AttributeReference = {
+    if (newQualifier.toSet == qualifier.toSet) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: no need to call `toSet` here?

**Comment by cloud-fan** [comment]
Overall LGTM, I think it's more clear to have `qualifier` as a `Option` instead of `Seq`.

**Comment by marmbrus** [comment]
LGTM pending tests.  The original signature is from very early on when I thought you could refer to a table both by name and alias.

**Comment by yhuai** [comment]
Thanks! Merging to master.

**Pull request closed without merging**